keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,350 @@
+  (although old replicas will be deleted afterwards) (see
+  [#40358](https://github.com/kubernetes/kubernetes/issues/40358))
+- if CA is not enabled, it's a deadlock since the replacement pod can't be
+  placed. The only workaround at this moment is to update app strategyType from
+  "RollingUpdate" to "Recreate".
+
+Neither of them is an ideal solution. A promising solution is to give user an
+option to trigger "toleration" mode when the cluster is out of resource. Then in
+aforementioned example, a third pod is "tolerated" to be put onto node1 (or
+node2). But keep it in mind, this behavior is only triggered upon resource
+shortage. For a 3-nodes cluster, the third pod will still be placed onto node3
+(if node3 is capable).
+
+### Risks and Mitigations
+
+Along with this feature, additional inevitable cost will be applied for each
+cycle of pod scheduling. So to mitigate potential performance impact, initial
+implementation will _probably_ limit the semantics of "even spreading" on
+`kubernetes.io/hostname`.
+
+We also need to make sure that our implementation will not have any performance
+penalty for pods that do not use this feature.
+
+## Design Details
+
+Basically there are two ways in API design and implementation:
+
+1. implemented as a "sub feature" inside Affinity and hence new API applies to
+   `pod.spec.affinity`. (referred as [Design A](#design-b) in this doc)
+1. implemented as a standalone feature and new API applies to `pod.spec`.
+   (referred as [Design B](#design-b) in this doc)
+
+### Design A
+
+A new structure called `EvenSpreading` is introduced and it's only effective
+when it's not nil.
+
+```go
+type EvenSpreading struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+}
+```
+
+In this design, `EvenSpreading` is a field in Affinity specs:
+
+```go
+type NodeAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+
+type PodAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+
+type PodAntiAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+```
+
+### Design B
+
+Unlike Design A, `EvenSpreading` acts as a standalone spec and applies to
+`pod.spec`. Similarly it's only effective when it's not nil.
+
+```go
+type PodSpec struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+```
+
+Inside `EvenSpreading`, we need hard affinityTerms (similar with
+`PodAffinityTerm`) and soft affinityTerms (similar with
+`WeightedPodAffinityTerm`). This describes when we perform even distribution,
+which pods are considered as a group.
+
+```go
+type EvenSpreading struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
```

Comment:
if there is only one topology key, i still fail to understand how is this different from the current pod affinity and anti affinity feature ?

Reply:
We can read `TopologyKey` as a (global) partition key. It's not the same as the individual "topologyKey" defined in "podAffinityTerm".

So the algorithm is like this:

1. we count which pods belong to the same "group" - determined by `[] podAffinityTerms`.
2. we also need to calculate by given the (global) `TopologyKey`, how the pods are distributed (e.g. 3/2/1)
3. then we can come up with the candidates node list which obey `MaxSkew`.

BTW: I agree the (global) `TopologyKey` may leads people to think of the "topologyKey" defined in "podAffinityTerm". So I [proposed earlier](https://github.com/kubernetes/enhancements/pull/851#discussion_r265261788) to rename to `DistributionKey` or `DistributedBy`.

Reply:
It sounds like global topology key is a combination of many topology domains like rack, server, chassic or other combinations. In case of even spreading across multiple domains, having the user come up with a global key is cryptic or i am not following along.  An example of what this global key would be , especially when multiple topology keys are in play would be helpful. We can also say that alpha, is only a single key. But then i would still allow the scope for multiple keys by making this a slice

Reply:
Things are easy if we only support one podAffinityTerm, hence one topologyKey,  which acts as the global topologyKey as well.

When `[]podAffinityTerms` come in, which imply there could be multiple topology keys, in this case, it's also doable to define **just one global key**. For example, incoming pod's 2 podAffinityTerms defines to 1) co-exist with pods which have label `foo` in the same "region", and also 2) co-exist with pods which have label `bar` in the same "zone". Per my explanation in https://github.com/kubernetes/enhancements/pull/851#discussion_r268054934, `[]podAffinityTerms` will deduce to a "node candidate list" - it's a node list, we don't care whatever region or zone the node is, yet. Then by giving a global key (it can be region/zone, or can be anything like rack/node), the candidate node list is then partitioned into "groups". And b/c it's doable to know how many pods belong to each "group", hence we can know which "group"(s) obey maxSkew, and eventually which nodes are good fits.

Reply:
i guess for the reader not familiar with podAffinityTerms, it would be good to include an example of what does that mean

Reply:
Also in case of multiple podAffinityTerms, i think what you are saying is each podAffinityTerm will produce a single node candidate list. We will need to take an intersection of the nodes list from all podAffinityTerms. Then the final candidate list is partitioned into groups based on combination of topoKey from each term. Does that blend with what you are suggesting ?

Reply:
Yeap, `[]podAffinityTerms` can be an option is because (1) it's an existing API and (2) it can express the semantics of almost all the cases.

But the concern is its complexity.

You can put comments on the latest doc in section "API options".